Extend and refactor `example-queries` command, rename to `benchmark-queries`

tanmay-9 opened the pull request:
1. Support input file (with SPARQL queries and a description for each) in both TSV and YML; refactor the parsing code accordingly
2. Add option to generate a YML result file suitable for processing with our evaluation web app; see #171 
3. Add unit tests
4. Rename from `example-queries` to `benchmark-queries` because that is really what this command is doing; the old functionality, which was a special case, is still there using the `--example-queries` option

hannahbast commented:
@tanmay-9 I tried out the latest version and noticed the following issues:

1. In src/qlever/commands/example-queries.py you iterate `for binding in results_json["bindings"]` and then `for obj in binding.values()`. But a `binding` may be `null` in the JSON (and then `None` in Python), in which case your second loop throws an error.

2. The current https://qlever.cs.uni-freiburg.de/evaluation throws an error at `engines-comparison.js:237:37` (cannot read `runtime_info`) when comparing results for Wikidata. Any idea what is going wrong there?

3. It seems that `qlever example-queries ... --generate-output-file` writes the same file in two locations: in `src/qlever/evaluation/output` and in the current directory. If that is so, I don't think it's a good idea because it's never a good idea to have the exact same file twice in different locations. If anything, the latter should be a symbolic link to the former.

4. Currently, `qlever example-queries` with output file generation requires *three* options, namely `--generate-output-file`, `--output-basename`, and `--backend-name`. This should, and in my view can, be a single option, for example, `--results-file <file name>`. I am aware that you use the current `--backend-name` option also for determining whether the engine is QLever, but this could also be done by convention by checking whether `qlever` occurs in the results file name.

tanmay-9 commented:
1. I have added a check for this in `example-queries` so that it doesn't fail.
2. I have added a lot of new changes to engine_comparison.js to make the compare results table generation more robust and less crash prone.
3. I have added the symlink functionality to `example-queries`
4. I agree that it would be better as a single option `--results-file`. The problem with that currently is that the evaluation web app is extracting the names of the datasets and engines automatically by parsing the filenames in the output folder which is guaranteed by the `--output-basename` and `--backend-name` options. If we remove these, then we would need to think of another way to consistently extract the dataset and engine names or ask the user to maintain filename consistency when using the `--results-file` option.

hannahbast commented:
@tanmay-9 Thanks a lot for the changes. One minor issue that is hopefully easy to fix: Right now, a "warning sign" is shown when the result size for one engine differs from the result size of the majority. However, for many of our benchmarks, the queries have only one result, which is a count. It would be very useful, if for those queries the warning sign would also appear if for one engine that count is different fron the majority.

hannahbast commented on src/qlever/commands/example_queries.py:
```diff
@@ -254,17 +387,15 @@ def execute(self, args) -> bool:
-            example_query_lines = example_query_lines.splitlines()
-        except Exception as e:
-            log.error(f"Failed to get example queries: {e}")
+        # Get the example queries either from queries_file or get_queries_cmd
+        example_query_lines = (
+            self.get_example_queries(get_queries_cmd=get_queries_cmd)
+            if args.queries_file is None
+            else self.get_example_queries(queries_file=args.queries_file)
+        )
+
+        if len(example_query_lines) == 0:
+            log.error("No example queries matching the criteria found")
```
This is no longer just "example" queries, see above. Rename appropriately

hannahbast commented on src/qlever/commands/example_queries.py:
```diff
@@ -278,6 +409,7 @@ def execute(self, args) -> bool:
         # processing time (seconds).
         query_times = []
         result_sizes = []
+        yaml_records = {"queries": []}
```
Give a better name, YAML is just the format and not the main thing this variable is about. Maybe `query_results`

hannahbast commented on src/qlever/commands/example_queries.py:
```diff
@@ -431,6 +573,16 @@ def execute(self, args) -> bool:
                         )
+                    elif accept_header == "application/qlever-results+json":
+                        try:
+                            # sed cmd to get the number between 2nd and 3rd double_quotes
+                            result_size = run_command(
+                                f"jq '.res[0]' {result_file}"
+                                " | sed 's/[^0-9]*\\([0-9]*\\).*/\\1/'",
+                                return_output=True,
+                            )
```
Shouldn't there be one function for extracting the single count from the result if there is one? Right now there is special code for this spread over several places

hannahbast commented on src/qlever/commands/example_queries.py:
```diff
@@ -493,9 +637,7 @@ def execute(self, args) -> bool:
-            # Remove the result file (unless in debug mode).
-            if args.log_level != "DEBUG":
-                Path(result_file).unlink(missing_ok=True)
+            error_msg_for_yaml = {}
```
Does the `_yml` make sense in the variable name here?

hannahbast commented on src/qlever/commands/example_queries.py:
```diff
@@ -552,10 +696,48 @@ def execute(self, args) -> bool:
         assert len(result_sizes) == len(query_times)
         assert len(query_times) + num_failed == len(example_query_lines)
 
+        if len(yaml_records["queries"]) != 0:
+            outfile_name = f"{dataset}.{engine}.results.yaml"
+            outfile = Path(args.results_dir) / outfile_name
+            self.write_query_data_to_yaml(
+                query_data=yaml_records,
+                out_file=outfile,
+            )
```
Shouldn't there be an error message if there are no query results?

hannahbast commented on src/qlever/commands/example_queries.py:
```diff
@@ -603,3 +785,125 @@ def execute(self, args) -> bool:
         # Return success (has nothing to do with how many queries failed).
         return True
+
+    def get_record_for_yaml(
+        self,
+        query: str,
+        sparql: str,
+        client_time: float,
+        result: str | dict[str, str],
+        result_size: int | None,
+        accept_header: str,
+    ) -> dict[str, Any]:
+        """
+        Construct a dictionary with query information for yaml file
```
This function should have a better name and comment. In particular, the input file can be a YML as well, so `yaml` is not a unique reference

hannahbast commented on src/qlever/commands/example_queries.py:
```diff
@@ -1,19 +1,28 @@
 from qlever.log import log, mute_log
 from qlever.util import run_command, run_curl_command
 
+MAX_RESULT_SIZE = 20
```
Should be an argument, e.g. `max-results-output-file`

hannahbast commented on src/qlever/commands/example_queries.py:
```diff
@@ -603,3 +785,125 @@ def execute(self, args) -> bool:
+        record["results"] = results
+        return record
+
+    def get_query_results(
+        self, result_file: str, result_size: int, accept_header: str
+    ) -> tuple[list[str], list[list[str]]]:
+        """
+        Return headers and results as a tuple
```
Improve name and comment. Since we have rather many helper functions now, they should have good and consistent names and clear comments (which does not mean long comments)

hannahbast commented on src/qlever/commands/example_queries.py:
```diff
@@ -603,3 +785,125 @@ def execute(self, args) -> bool:
+                results.append(result)
+            return results_json["headers"], results
+        else:  # text/turtle
+            graph = Graph()
+            graph.parse(result_file, format="turtle")
```
Make prefix `rdflib` explicit

hannahbast commented on src/qlever/commands/example_queries.py:
```diff
@@ -603,3 +785,125 @@ def execute(self, args) -> bool:
+                results.append([str(s), str(p), str(o)])
+            return headers, results
+
+    @staticmethod
+    def write_query_data_to_yaml(
+        query_data: dict[str, list[dict[str, Any]]], out_file: Path
+    ) -> None:
+        """
+        Write yaml record for all queries to output yaml file
+        """
```
Dito (regarding name and comment)

hannahbast commented on src/qlever/commands/example_queries.py:
```diff
@@ -212,89 +227,245 @@ def sparql_query_type(self, query: str) -> str:
+        get_queries_cmd += f" | sed -n '{sed_arg}'"
+        if query_regex:
+            get_queries_cmd += f" | grep -Pi {shlex.quote(query_regex)}"
+        return get_queries_cmd
```
We discussed that there should be an additional option `--example-queries´ (and the whole command should be renamed to `benchmark-queries`, so that the code looks them same no matter whether you parse from TSV or YAML (just the parsing is different). The user should then *either* specify `--example-queries` or one of the `queries-...`, and there should be an informative error message otherwise

hannahbast commented on src/qlever/commands/example_queries.py:
```diff
@@ -212,89 +227,245 @@ def sparql_query_type(self, query: str) -> str:
 
-    def get_example_queries(
+        try:
+            tsv_queries = []
+            for query_idx in query_indices:
+                if query_idx >= total_queries:
+                    log.error(
+                        "Make sure --query-ids don't exceed the total "
+                        "queries in the YML file"
+                    )
+                    return []
+                query = data["queries"][query_idx]
+
+                # Only include queries that match the query_regex if present
+                if query_regex:
+                    pattern = re.compile(query_regex, re.IGNORECASE)
+                    if not any(
+                        [
+                            pattern.search(query["query"]),
+                            pattern.search(query["sparql"]),
+                        ]
+                    ):
+                        continue
+
+                tsv_queries.append(f"{query['query']}\t{query['sparql']}")
```
This code should come after the parsing, to avoid code duplication. The two parsers provide the same intermediate representation of the queries, and that one can then filter based on `query-ids` and `query-regex`.

hannahbast commented on src/qlever/commands/example_queries.py:
```diff
@@ -212,89 +227,245 @@ def sparql_query_type(self, query: str) -> str:
+                        return_output=True,
+                    )
+                except Exception as e:
+                    error_msg = get_json_error_msg(e)
```
Looks like this should be a separate function + it should be tested.

hannahbast commented on src/qlever/commands/example_queries.py:
```diff
@@ -212,89 +227,245 @@ def sparql_query_type(self, query: str) -> str:
+
+                # CASE 2: Downloading the full result (TSV, CSV, Turtle, JSON).
+        else:
+            if (
+                accept_header == "text/tab-separated-values"
+                or accept_header == "text/csv"
+            ):
+                result_size = run_command(
+                    f"sed 1d {result_file} | wc -l", return_output=True
+                )
+            elif accept_header == "text/turtle":
+                result_size = run_command(
+                    f"sed '1d;/^@prefix/d;/^\\s*$/d' {result_file} | wc -l",
+                    return_output=True,
+                )
+            elif accept_header == "application/qlever-results+json":
+                result_size = run_command(
+                    f'jq -r ".resultsize" {result_file}',
+                    return_output=True,
+                )
+            else:
+                try:
+                    result_size = int(
+                        run_command(
+                            f'jq -r ".results.bindings | length"'
+                            f" {result_file}",
+                            return_output=True,
+                        ).rstrip()
+                    )
+                except Exception as e:
+                    error_msg = get_json_error_msg(e)
+                if result_size == 1:
+                    try:
+                        single_int_result = int(
+                            run_command(
+                                f'jq -e -r ".results.bindings[0][] | .value"'
+                                f" {result_file}",
+                                return_output=True,
+                            ).rstrip()
+                        )
```
Looks like this should be a separate function + it should be tested.

hannahbast commented on src/qlever/commands/benchmark_queries.py:
```diff
@@ -93,6 +93,15 @@ def additional_arguments(self, subparser) -> None:
+            action="store_true",
+            default=False,
+            help=(
+                "Run the example-queries for the given --ui-config "
```
```suggestion
                "Run the example queries for the given --ui-config "
```

hannahbast commented on src/qlever/commands/benchmark_queries.py:
```diff
@@ -93,6 +93,15 @@ def additional_arguments(self, subparser) -> None:
+            default=False,
+            help=(
+                "Run the example-queries for the given --ui-config "
+                "instead of the benchmark queries from a tsv/yml file"
```
```suggestion
                "instead of the benchmark queries from a TSV or YML file"
```

hannahbast commented on src/qlever/commands/benchmark_queries.py:
```diff
@@ -271,9 +291,18 @@ def parse_queries_tsv(
-    ) -> list[str]:
+    def parse_queries_tsv(queries_file: str) -> list[str]:
+        """
+        Parse the queries_tsv file and return a list of tab-separated queries
+        (query_description, full_sparql_query)
```
The internal representation should be more abstract, like a list of tuples.

hannahbast commented on src/qlever/commands/benchmark_queries.py:
```diff
@@ -227,41 +235,53 @@ def sparql_query_type(self, query: str) -> str:
-        Construct get_queries_cmd from queries_tsv file if present or use
-        example queries by using ui_config. Use query_ids and query_regex to
-        filter the queries
+        Given a tab-separated list of queries, filter them and keep the
+        ones which are a part of query_ids or match with query_regex
```
The input should be a list of tuples, see below. Change name to `filter_queries`

hannahbast commented on src/qlever/commands/benchmark_queries.py:
```diff
@@ -227,41 +235,53 @@ def sparql_query_type(self, query: str) -> str:
-        Parse the queries_tsv file (or example queries cmd if file absent)
-        and return a list of tab-separated queries
-        (query_description, full_sparql_query)
+        Execute the given bash command to fetch tsv queries and return a
+        list of tab-separated queries (query_description, full_sparql_query)
```
List of tuples as well

hannahbast commented on src/qlever/commands/benchmark_queries.py:
```diff
@@ -439,18 +431,27 @@ def get_json_error_msg(e: Exception) -> dict[str, str]:
+        """
+        When downloading the full result of a query with accept header as
+        application/sparql-results+json and result_size == 1, get the single
+        integer result value (if any)
```
```suggestion
        integer result value (if any).
```

hannahbast reviewed (changes requested):
1-1 with Tanmay, round 3. Some minor comments left. The web app should be factored out of this. Thanks a lot.
